Limited external connection warning incorrect

theskinnyghost opened the pull request:
### Description of the Change
Closes #1071 

### How to test the Change

1. Go to the WP Dashboard > Distributor > External Connections > Add new
2. Click Manually Set Up Connection.
3. Enter an incorrect username and password combination
4. Enter a valid REST API endpoint in the URL field for a site that is running Distributor
5. Instead of the yellow warning, you should see a red error with a `Authentication failed due to insufficient or invalid credentials.` message instead
6. The push and pull permissions list will show no post types can be pushed or pulled
7. Save the connection
8. Visit Distributor > Pull Content
9. The new connection will not be listed as the site does not have any pullable content types
10. Publish a new post
11. Click "Distribute"
12. The new external connection will not be listed as the site does not have any pushable content types

### Changelog Entry

> Changed external connection status error messages when checking status

### Credits
Props @theskinnyghost 

### Checklist:
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

theskinnyghost committed:
Show error instead of warning when incorrect credentials are used for external connections

theskinnyghost commented:
@peterwilsoncc I tried to run the cypress tests locally but wasn't able to so I'm not sure if I need to make any additional setup configuration to run them properly. Here's the error that I get when I run `npm run cypress:run` 
![image](https://github.com/10up/distributor/assets/8986630/c5931991-3ee6-4856-be18-baccee4f2da5)

Also, is this the test I would need to update?
https://github.com/10up/distributor/blob/36d1c76/tests/cypress/e2e/external-connection.test.js#L99

peterwilsoncc commented on assets/js/admin-external-connection.js:
```diff
@@ -348,6 +348,18 @@ function checkConnections() {
+					'distributor'
+				);
+
+				speak(
+					__( 'No connection found.', 'distributor' ),
+					'polite'
+				);
```
Coding standards

```suggestion
				speak( __( 'No connection found.', 'distributor' ), 'polite' );
```

peterwilsoncc reviewed (commented):
Thanks for the PR, Luca.

This is my fault but I the rest-api endpoint needs an update in the function `check_post_types_permissions()`, the permission check loop should be:

```php
foreach ( $types as $type ) {
	$caps = $type->cap;

	if ( current_user_can( $caps->edit_posts ) ) {
		$response['can_post'][] = $type->name;
	}

	if ( current_user_can( $caps->edit_posts ) && current_user_can( $caps->create_posts ) && current_user_can( $caps->publish_posts ) ) {
		$response['can_post'][] = $type->name;
	}
}
```

While testing, I noticed that if Distributor is not installed it shows the old limited connection message, as in this screenshot. This too will need to be updated to an error.

![Screen Shot 2023-10-03 at 10 46 08 am](https://github.com/10up/distributor/assets/519727/6015052f-e05a-49fd-a4de-a75fd4babdc4)


You're correct, the test you've linked to in `external-connection.test.js` is the one that needs an update. 

Getting Cypress running locally can be a little fiddly, were you able to get the environment running in Docker or did it fail before then?

theskinnyghost committed:
Error when distributor is not available on external connection

theskinnyghost committed:
Update check post types permissions logic

theskinnyghost committed:
Update tests to match error for external connection

theskinnyghost commented:
Hey @peterwilsoncc, thanks for your feedback.

I made an update to show an error for the use case you mentioned above (when distributor is not installed).

I also updated the `check_post_types_permissions()` logic as per your suggestion but I'm not sure I did it right so I left a comment for you to check.

RE: Cypress Tests – I only ran `npm run cypress:run` and that's the error I got. Do I need to run additional commands? I use Local Flywheel not Docker.

theskinnyghost commented on includes/rest-api.php:
```diff
@@ -600,8 +600,11 @@ function check_post_types_permissions() {
+		$caps = $type->cap;
+
+		if ( current_user_can( $caps->edit_posts ) ) {
+			$response['can_post'][] = $type->name;
```
@peterwilsoncc should this be `$response['can_get'][] = $type->name;` instead?

peterwilsoncc replied:
> should this be `$response['can_get'][] = $type->name;` instead?

Yes, it should be. Sorry about that and thanks for picking up my silly moment.

theskinnyghost replied:
Thank you and no worries. Updated! 😄

theskinnyghost committed:
Fix typo in check_post_types_permissions()

peterwilsoncc commented:
> RE: Cypress Tests – I only ran `npm run cypress:run` and that's the error I got. Do I need to run additional commands? I use Local Flywheel not Docker.

The E2E tests use [`@wordpress/env`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/) which requires Docker. You may be able to change the URL returned by this function to use local but I've never had much luck doing that as a vagrant user.

https://github.com/10up/distributor/blob/0387c94509f4dbde5781628671b0b58abf5e3884/tests/cypress/config.js#L29-L48

I can take a look at the changes required.

peterwilsoncc commented on assets/js/admin-external-connection.js:
```diff
@@ -314,7 +314,11 @@ function checkConnections() {
+			} else if (
+				response.data.errors.no_external_connection ||
+				'no' === response.data.is_authenticated ||
+				response.data.errors.no_distributor
```
```suggestion
				response.data.errors.no_distributor ||
				! response.data.can_post.length
```

peterwilsoncc commented on assets/js/admin-external-connection.js:
```diff
@@ -337,6 +341,18 @@ function checkConnections() {
+						),
+						'polite'
+					);
 				} else {
```
This is to make it a little clearer that the site has Distributor but the account has low or no permissions.

```suggestion
				} else if (
					'no' === response.data.is_authenticated ||
					response.data.errors.no_types
				) {
					endpointResult.innerText = __(
						'Authentication failed due to insufficient or invalid credentials.',
						'distributor'
					);
					speak(
						__(
							'Authentication failed due to insufficient or invalid credentials.',
							'distributor'
						),
						'polite'
					);
				} else {
```

peterwilsoncc commented on tests/cypress/e2e/external-connection.test.js:
```diff
@@ -109,9 +109,9 @@ describe( 'Admin can add a new external connection', () => {
+		cy.get( '.endpoint-result' ).should(
 			'contain.text',
-			'Limited connection established.'
+			'No connection found'
```
```suggestion
			'Authentication failed due to insufficient or invalid credentials.'
```

peterwilsoncc reviewed (commented):
I retested and was still seeing some unexpected warnings rather than errors, so have added a few more suggestions.

With these changes, I think we'll then be able to remove the `else if` beginning on [line 367](https://github.com/10up/distributor/blob/a7ec0387d89878ceff62c85d5aecedd31e9bb2e3/assets/js/admin-external-connection.js#L367-L411) -- it's outside the diff so I'm unable to include it as an inline suggestion.

theskinnyghost committed:
Apply suggestions from code review

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

theskinnyghost committed:
Merge branch 'develop' into issue-1071

theskinnyghost commented:
@peterwilsoncc thanks again for your feedback. I implemented the changes you suggested but wasn't totally sure we could remove the `else if` from line 367 as that would completely remove warnings. Is that what you'd like to achieve here?

re: tests - I was able to run the tests this time by running `npm env:start` before running`npm run cypress:run`. However, a lot of tests are failing (even when I use the `develop` or `trunk` branch). I wonder if I'm missing anything here?

peterwilsoncc commented:
> wasn't totally sure we could remove the `else if` from line 367 as that would completely remove warnings. Is that what you'd like to achieve here?

Yes, that's what I was after. The two conditions `response.data.errors.no_distributor || ! response.data.can_post.length` are both included in the first block now and throw errors instead of warnings.

A downside to switching to the dedicated Distributor endpoint is that it's basically all-or-nothing for connections at the moment. They'll either error or succeed, we're no longer able to have the partial connections that used to trigger warnings. 

As the block-editor becomes more important, the partial connections of old become less effective as the block editor requires access to the raw content.

> re: tests - I was able to run the tests this time by running `npm env:start` before running`npm run cypress:run`. However, a lot of tests are failing (even when I use the `develop` or `trunk` branch). I wonder if I'm missing anything here?

I'll test from scratch locally and reach out via the 10up Slack. It's possible the docs need to be updated.

theskinnyghost committed:
Remove warnings when setting up an external connection

theskinnyghost committed:
Remove 'warning' errors throughout the code

theskinnyghost commented:
Thanks @peterwilsoncc, I've updated the code and removed the warnings since they aren't applicable anymore. I also updated the tests and removed the test for warnings.
I also think we should add an additional test to make sure the `Distributor not installed on remote site.` error appears when that's the case but I'm not sure how to set this up with the current setup and also wanted to double check with you if that's needed.
I was able to run the tests just fine following the instructions you provided on 10up's Slack, thanks! 😄

peterwilsoncc commented:
For others, these are the steps currently required to get the commands required to get e2e tests running locally.

Please ensure Docker is running before running these commands.

```
npm run env:destroy // Just to get back to scratch
npm run env:start
npm run copy-htaccess
npm run cypress:run
```

peterwilsoncc reviewed (approved):
Thanks Luca, this looks good to me.

Thanks for your patience during the review process, quite a bit had changed.
